gazebo/Server.cc

```diff
@@ -542,6 +548,110 @@ void Server::Fini()
+    performanceMetricsMsg.set_real_time_factor(simTime.Double());
+  else
+    performanceMetricsMsg.set_real_time_factor(0.0);
+
+  /// update sim time for sensors
+  for (auto model: world->Models())
+  {
+    for (auto link: model->GetLinks())
+    {
+      for (unsigned int i = 0; i < link->GetSensorCount(); i++)
+      {
+        std::string name = link->GetSensorName(i);
+        sensors::SensorPtr sensor = sensors::get_sensor(name);
+
+        auto ret = sensorsLastMeasurementTime.insert(
+            std::pair<std::string, gazebo::common::Time>(name, 0));
+        worldLastMeasurementTime.insert(
+                std::pair<std::string, gazebo::common::Time>(name, 0));
+        if (ret.second==false)
+        {
+          double updateSimRate =
+            (sensor->LastMeasurementTime() - sensorsLastMeasurementTime[name]).Double();
+          double updateRealRate =
+            (world->RealTime() - worldLastMeasurementTime[name]).Double();
+          sensorsLastMeasurementTime[name] = sensor->LastMeasurementTime();
+          if (updateSimRate > 0.0)
+          {
+            struct sensorPerformanceMetricsType emptySensorPerfomanceMetrics;
+            auto ret2 = sensorPerformanceMetrics.insert(
+                std::pair<std::string, struct sensorPerformanceMetricsType>
+                  (name, emptySensorPerfomanceMetrics));
+            if (ret2.second==false)
+            {
+              sensorPerformanceMetrics[name].sensorSimUpdateRate = 1.0/updateSimRate;
+              sensorPerformanceMetrics[name].sensorRealUpdateRate = 1.0/updateRealRate;
+              worldLastMeasurementTime[name] = world->RealTime();
+
+              // Special case for stereo cameras
+              sensors::CameraSensorPtr cameraSensor =
+                std::dynamic_pointer_cast<sensors::CameraSensor>(sensor);
+              if (nullptr != cameraSensor)
+              {
+                sensorPerformanceMetrics[name].sensorFPS = cameraSensor->Camera()->AvgFPS();
+              }
+              else
+              {
+                sensorPerformanceMetrics[name].sensorFPS = -1;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  for (auto sensorPerformanceMetric: sensorPerformanceMetrics)
+  {
+    msgs::PerformanceMetrics::PerformanceSensorMetrics * performanceSensorMetricsMsg =
+      performanceMetricsMsg.add_sensor();
+    performanceSensorMetricsMsg->set_sensor_name(sensorPerformanceMetric.first);
+    performanceSensorMetricsMsg->set_real_sensor_update_rate(
+      sensorPerformanceMetric.second.sensorRealUpdateRate);
+    performanceSensorMetricsMsg->set_sim_sensor_update_rate(
+      sensorPerformanceMetric.second.sensorSimUpdateRate);
+    if (sensorPerformanceMetric.second.sensorFPS >= 0.0)
+    {
+      performanceSensorMetricsMsg->set_fps(
+        sensorPerformanceMetric.second.sensorFPS);
+    }
+  }
+
+  // Publish data
+  if (performanceMetricsPub && performanceMetricsPub->HasConnections())
```

Comment:
can we move this check to the beginning of the function and return if there are not connections? The down side is that we get one spike in values the first time we publish a msg.